pyocd/rtos/__init__.py

```diff
@@ -15,14 +15,17 @@
  limitations under the License.
 """
 
+from collections import OrderedDict
 from .argon import ArgonThreadProvider
+from .bare_arm_m import BareARMMThreadProvider
 from .freertos import FreeRTOSThreadProvider
 from .zephyr import ZephyrThreadProvider
 from .rtx5 import RTX5ThreadProvider
 
-RTOS = {
-          'Argon' : ArgonThreadProvider,
-          'FreeRTOS' : FreeRTOSThreadProvider,
-          'Zephyr' : ZephyrThreadProvider,
-          'RTX5' : RTX5ThreadProvider,
-         }
+RTOS = OrderedDict([
```

Comment:
I think I'd rather have special logic for the Bare ARM-M provider so that it gets used as a fallback instead of requiring the RTOS list to be sorted. In other words, Bare ARM-M would not be in the `RTOS` dict.

Reply:
It would also be nice to use Bare ARM-M even when another provider is selected but not yet active, for instance in `main()` before the kernel has started running. While this isn't really applicable to Mbed OS, it would be useful for the other RTOSes and RTX5 outside of Mbed.

Reply:
Agree - this is certainly one of the "edge cases" I was thinking about.

I think this is doable - my current thought is that the GDBServer will have two thread providers in use - 1 core one (always present) and 1 RTOS one.  The RTOS one gets the opportunity to override the core one if it decides the RTOS is active.

Will start working in this direction.